src/substrait/function_registry.py

```diff
@@ -0,0 +1,378 @@
+from substrait.gen.proto.parameterized_types_pb2 import ParameterizedType
+from substrait.gen.proto.type_pb2 import Type
+from importlib.resources import files as importlib_files
+import itertools
+from collections import defaultdict
+from collections.abc import Mapping
+from pathlib import Path
+from typing import Any, Optional, Union
+from .derivation_expression import evaluate
+
+import yaml
+import re
+
+
+DEFAULT_URI_PREFIX = "https://github.com/substrait-io/substrait/blob/main/extensions"
+
+
+# mapping from argument types to shortened signature names: https://substrait.io/extensions/#function-signature-compound-names
+_normalized_key_names = {
+    "i8": "i8",
+    "i16": "i16",
+    "i32": "i32",
+    "i64": "i64",
+    "fp32": "fp32",
+    "fp64": "fp64",
+    "string": "str",
+    "binary": "vbin",
+    "boolean": "bool",
+    "timestamp": "ts",
+    "timestamp_tz": "tstz",
+    "date": "date",
+    "time": "time",
+    "interval_year": "iyear",
+    "interval_day": "iday",
+    "interval_compound": "icompound",
+    "uuid": "uuid",
+    "fixedchar": "fchar",
+    "varchar": "vchar",
+    "fixedbinary": "fbin",
+    "decimal": "dec",
+    "precision_time": "pt",
+    "precision_timestamp": "pts",
+    "precision_timestamp_tz": "ptstz",
+    "struct": "struct",
+    "list": "list",
+    "map": "map",
+}
+
+
+def normalize_substrait_type_names(typ: str) -> str:
+    # Strip type specifiers
+    typ = typ.split("<")[0]
+    # First strip nullability marker
+    typ = typ.strip("?").lower()
+
+    if typ.startswith("any"):
+        return "any"
+    elif typ.startswith("u!"):
+        return typ
+    else:
+        return _normalized_key_names[typ]
```

Comment:
What happens if the short name is used instead?  Consider throwing a specific exception here so it's more clear what an end user should do.

Reply:
Is using short names in extension yamls legal? I changed dict search to throw a specific exception.

Reply:
A quick glance of functions_arithmetic.yaml contains a lot of i8 and i16 arguments.  If you put together a test which verifies that you pass all of the core YAML extensions you should be good to go.

My general practice when implementing a spec is to be exact as possible on output and lenient on input.  But since we have the nearly full test set that should be enough.

Reply:
instantiating the registry loads all extensions by default, so they are all implicitly being tested. I'll work on more explicit tests, as well, though